libmscore/figuredbass.cpp

```diff
@@ -1770,6 +1773,256 @@ FiguredBass* Score::addFiguredBass()
       select(fb, SelectType::SINGLE, 0);
       return fb;
       }
+      
+//---------------------------------------------------------
+//
+// METHODS relating to realizing the figured bass
+//
+//
+//---------------------------------------------------------
+      
+void FiguredBass::realizeFiguredBass(Score *score) {
+      Q_ASSERT(score);
 
+      for (Measure* measure = score->firstMeasure(); measure; measure = measure->nextMeasure()) {
+            realizeFiguredBass(measure);
+      }
+}
+      
+void FiguredBass::realizeFiguredBass(Measure *measure) {
+      Q_ASSERT(measure);
+      const Segment::Type st = Segment::Type::ChordRest;
+      for (Segment* segment = measure->first(st); segment; segment = segment->next(st)) {
+            realizeFiguredBass(segment);
+      }
+}
+      
+void FiguredBass::realizeFiguredBass(Segment *segment) {
+      Q_ASSERT(segment);
+      int etrack = segment->score()->nstaves() * VOICES;
+
+      for (int track = 0; track < etrack; ++track) {
+            Element *element = segment->element(track);
+            if ( element == 0)
+                  continue;
+            if ( element->type() != Element::Type::CHORD)
+                  continue;
+            {
+                  vector<int> harmonys;
+                  Chord *chord = static_cast<Chord*>(element);
+                  if ( calcFiguredBass(chord, harmonys) ) { // appends entries into harmonys corresponding musical harmony pitches
+                        for (int pitch : harmonys) {
+                              NoteVal *nval = new NoteVal(pitch);
+                              chord->score()->addNote(chord,*nval);
+                        }
+                  } else {
+                        //qDebug("no figured bass on chord on track %d", chord->track());
+                  }
+            }
+      }
+}
+      
+// populate the given harmonys vector with the pitches of the harmony notes implied by the figured base on the given chord
+bool FiguredBass::calcFiguredBass(Chord *chord, vector<int> & harmonys) {
+      Q_ASSERT(harmonys.size() == 0);
+      Segment *segment = chord->segment();
+      int octave = 0; // 0, or 12 if we want the harmony 1 octave up
+      FiguredBassItem::Modifier none = FiguredBassItem::Modifier::NONE;
+      if (find_if(segment->annotations().cbegin(),
+                  segment->annotations().cend(),
+                  [chord](Element *e) {
+                        return ( e->type() == Element::Type::FIGURED_BASS
+                                && e->track() == chord->track());
+                  }) == segment->annotations().cend())
+      {
+             return false; // no figured base to annotate
+      }
+      
+      if ( chord->notes().size() == 0 ) {
+            qDebug("no notes on chord");
+            return false; // no notes in the chord, not sure if this is possible?
```

Comment:
I do not think a chord can have no notes, but a Continuo indication may occur on rests (not very frequent, but it may happen, I have met this in early XVIII c. French music).